### PR TITLE
Splitting sandpack from main bundle

### PR DESCRIPTION
**Pull request by harish-sethuraman**
Tried to split the sandpack and added a fallback with suspense

Adding suspense to `Sandpack` reduced `~9.55KB`
Adding a wrapper to `codeBlock` and `Sandpack` reduced `~150KB`

Fallback preview here: https://beta-reactjs-org-namamgdvj-fbopensource.vercel.app/learn

### CODE REVIEW

**Comment by harish-sethuraman** [comment]
Sandpack (without highlighting and output) renders, Then the fallback and then the complete sandpack... 

https://user-images.githubusercontent.com/32865581/151690964-75ac2121-0817-4ff6-946e-4e2542910ec7.mov

File: beta/src/components/MDX/MDXComponents.tsx [hunk 1]
```diff
@@ -3,7 +3,7 @@
  */
 
 import * as React from 'react';
-
+import {SandpackSetup} from '@codesandbox/sandpack-react';
```
Comment (gaearon) [inline beta/src/components/MDX/MDXComponents.tsx:6]: this should just be `import type`?

Comment (harish-sethuraman) [reply beta/src/components/MDX/MDXComponents.tsx:6]: Yes will change

**Comment by gaearon** [comment]
Prerequisite fixes for this to work:

- [ ] https://github.com/vercel/next.js/pull/33861
- [ ] https://github.com/mdx-js/mdx/pull/1924

File: beta/src/components/MDX/CodeBlock/index.tsx [hunk 2]
```diff
@@ -2,6 +2,30 @@
...
+import * as React from 'react';
+const CodeBlock = React.lazy(
+  () =>
+    import('./CodeBlock').then(
+      (x) =>
+        new Promise((resolve) => {
+          setTimeout(() => resolve(x), 3000);
+        })
+    ),
+  {
```
[the remark this answers is not in the document]
Comment (gaearon) [reply beta/src/components/MDX/CodeBlock/index.tsx:14]: yeah this isn't needed

**Comment by gaearon** [comment]
Next thing I'm seeing is that with production builds, HTML always includes fallback rather than content. I suspect this might just be some Next.js optimization or because we haven't enabled the `concurrentFeatures` flag. Maybe it's due to static generation.

**Comment by gaearon** [comment]
OK, i can confirm this is because SSG skips content and only shows fallbacks. Not sure what Next.js canonical solution is. I'll ask them.

**Comment by gaearon** [comment]
I guess that doesn't really matter though because we should just make really good fallbacks with pre tags. Then this is what we want to show on SSR anyway. So that's the next step.

**Comment by harish-sethuraman** [comment]
I've added fallback but actually at some point when the syntax highlighting is getting applied there is minimal shift.. 
This appears in the main branch too...

Screenshot 2022-02-01 at 7 45 38 PM

**Comment by danilowoz** [comment]
Hey guys, FYI, Sandpack will be able to provide a syntax highlight component on SSR soon, which will avoid any flash between server and client, more details in this [PR](https://github.com/codesandbox/sandpack/pull/276). Otherwise, I think it makes sense to only Suspense the code-editor component (which contains the heaviest dependencies), instead of the whole Sandpack package, like https://github.com/reactjs/reactjs.org/pull/4112.

Let me know if there is anything we can do from our side to make it easier.

**Comment by gaearon** [comment]
I think @danilowoz means we'd still want to lazy-load `SandpackCodeViewer`, not sure though

**Comment by harish-sethuraman** [comment]
> I think @danilowoz means we'd still want to lazy-load `SandpackCodeViewer`, not sure though

We are anyway splitting `codeBlock` component that consumes `SandpackCodeViewer` right?

**Comment by danilowoz** [comment]
yeah, I'd narrow it down and lazily load the following components:
- CodeBlock: https://github.com/reactjs/reactjs.org/blob/main/beta/src/components/MDX/CodeBlock/CodeBlock.tsx#L93-L97
- Sandpack: https://github.com/reactjs/reactjs.org/blob/main/beta/src/components/MDX/Sandpack/CustomPreset.tsx#L73-L81

Both contain the same dependencies and most of the heaviest parts from CodeMirror.

**Comment by harish-sethuraman** [comment]
We have lazily loaded them both 😄

**Comment by gaearon** [comment]
filed https://github.com/vercel/next.js/issues/33879 for why fallbacks appear in SSG

**Comment by gaearon** [comment]
so the issue here was that we really shouldn't be doing this without the flag: https://github.com/vercel/next.js/issues/33879#issuecomment-1027324625

so we need to wait for https://github.com/vercel/next.js/issues/31317

**Comment by gaearon** [comment]
actually https://github.com/vercel/next.js/pull/33886 would be enough

**Comment by gaearon** [comment]
The padding on code blocks jumps a bit.

**Comment by harish-sethuraman** [comment]
Pretty weird before the preview before I reverse merged with main has perfect fallback without any jumps in the UI. 

Before: https://beta-reactjs-org-namamgdvj-fbopensource.vercel.app/reference/usestate

Now: https://beta-reactjs-org-o28nb0eep-fbopensource.vercel.app/reference/usestate

**Comment by gaearon** [comment]
maybe because we merged a sandpack bump so it's a new version

**Comment by harish-sethuraman** [comment]
> maybe because we merged a sandpack bump so it's a new version

Yes, It is caused because of the sandpack bump. Verified it in local by checking out before that change

**Comment by harish-sethuraman** [comment]
`v12.0.11-canary.9` with the changes we require broke the build 🤔.

**Comment by harish-sethuraman** [comment]
`v12.0.11-canary.10` even this doesn't work I even tried latest canary. Only the stable tag worked. Should we report this ?

**Comment by harish-sethuraman** [comment]
@danilowoz during the initial render the sandpack content is loaded but without the code formatting (it is loaded as string). Later we get the fallback and then the formatted code. Need help in finding why code is rendered as string during initial render.

PS: Build fails you can try the same change in local

wYuRiv-W

**Comment by danilowoz** [comment]
@harish-sethuraman currently, it's not possible, due to some CodeMirror limitations. But Sandpack is going to support it very soon, as we made some important progress in SSR support (https://github.com/codesandbox/sandpack/pull/276).

If you want to give it a try, take a look at this PR https://github.com/danilowoz/reactjs.org/pull/1 - but, please don't use it for production, there are a few breaking change comming in this implementation.

**Comment by harish-sethuraman** [comment]
This change splits sandpack and codeblock into separate chunks. So there are lot of components to test. 

- [x] Codeblocks (ex: https://beta-reactjs-org-ec9ekfllb-fbopensource.vercel.app/learn)
- [x] Code blocks with terminals  (ex: https://beta-reactjs-org-ec9ekfllb-fbopensource.vercel.app/apis)
- [x] Sandpack (ex: https://beta-reactjs-org-ec9ekfllb-fbopensource.vercel.app/learn/installation)
- [x] API anatomy (ex: https://beta-reactjs-org-ec9ekfllb-fbopensource.vercel.app/apis/usestate)

**Comment by gaearon** [comment]
looks great. wanna press merge? if something breaks we'll roll back

**Comment by harish-sethuraman** [comment]
🚀

**Pull request merged**